toolchain/check/handle_binding_pattern.cpp

```diff
@@ -216,45 +212,49 @@ static auto HandleAnyBindingPattern(Context& context, Parse::NodeId node_id,
+       .parent_scope_id = context.scope_stack().PeekNameScopeId()});
+
+  return HandleAnyBindingPattern(context, node_id, name_node, entity_name_id,
+                                 type_node, type_inst_id,
                                  Parse::NodeKind::LetBindingPattern);
 }
 
 auto HandleParseNode(Context& context, Parse::VarBindingPatternId node_id)
     -> bool {
-  return HandleAnyBindingPattern(context, node_id,
-                                 Parse::NodeKind::VarBindingPattern);
-}
+  auto [type_node, type_inst_id] = context.node_stack().PopExprWithNodeId();
 
-auto HandleParseNode(Context& context,
-                     Parse::CompileTimeBindingPatternStartId node_id) -> bool {
-  // Make a scope to contain the `.Self` facet value for use in the type of the
-  // compile time binding. This is popped when handling the
-  // CompileTimeBindingPatternId.
-  context.scope_stack().PushForSameRegion();
+  // `template` is incorrect here, but is diagnosed in parse.
+  context.node_stack()
+      .PopAndDiscardSoloNodeIdIf<Parse::NodeKind::TemplateBindingName>();
 
-  // The `.Self` must have a type of `FacetType`, so that it gets wrapped in
-  // `FacetAccessType` when used in a type position, such as in `U:! I(.Self)`.
-  // This allows substitution with other facet values without requiring an
-  // additional `FacetAccessType` to be inserted.
-  SemIR::FacetTypeId facet_type_id =
-      context.facet_types().Add(SemIR::FacetTypeInfo{});
-  auto const_id = EvalOrAddInst<SemIR::FacetType>(
-      context, node_id,
-      {.type_id = SemIR::TypeType::TypeId, .facet_type_id = facet_type_id});
-  auto type_id = context.types().GetTypeIdForTypeConstantId(const_id);
+  auto [name_node, name_id] = context.node_stack().PopNameWithNodeId();
+  auto entity_name_id = context.entity_names().Add(
+      {.name_id = name_id,
+       .parent_scope_id = context.scope_stack().PeekNameScopeId()});
 
-  MakePeriodSelfFacetValue(context, type_id);
-  return true;
+  return HandleAnyBindingPattern(context, node_id, name_node, entity_name_id,
+                                 type_node, type_inst_id,
+                                 Parse::NodeKind::VarBindingPattern);
 }
 
 auto HandleParseNode(Context& context,
-                     Parse::CompileTimeBindingPatternId node_id) -> bool {
-  // Pop the `.Self` facet value name introduced by the
-  // CompileTimeBindingPatternStart.
-  context.scope_stack().Pop();
+                     Parse::CompileTimeBindingPatternStartId node_id) -> bool {
+  auto is_template =
+      context.node_stack()
+          .PopAndDiscardSoloNodeIdIf<Parse::NodeKind::TemplateBindingName>();
+
+  auto name_id = context.node_stack().PeekNameId();
+  auto entity_name_id = SemIR::EntityNameId::None;
 
-  auto node_kind = Parse::NodeKind::CompileTimeBindingPattern;
   const DeclIntroducerState& introducer =
       context.decl_introducer_state_stack().innermost();
   if (introducer.kind == Lex::TokenKind::Let) {
```

Comment:
On rereading this I'm not sure I understand what's going on here: what other kinds of compile-time bindings are there besides `Let`? (I was first wondering if we could drop the ``"`let` "`` from the TODO diagnostic below, but maybe I'm missing something here.)